feat: Noports CLI implementation

srieteja opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines in CONTRIBUTING.md

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->
Closes https://github.com/atsign-foundation/noports/issues/1976

**- What I did**
- Introduced Noports CLI with `activate` and `issue-keys` commands
- Implemented a facade pattern for both commands and added individual fromArgs factories to make instantiation easier.
- `issue-keys` generates copy-pasteable activation strings, `activate` accepts them.
- Created a CLILoggingHandler to display cleaner logs for CLIs

**- How I did it**
#### Core Architecture:

- Created Activate class with CRAM and APKAM enrollment support
- Created IssueKeys class to generate enrollment strings and auto-approve requests
- Implemented NoPortsCommand enum for type-safe command routing in noports.dart

#### Activate Command:

- Added noports activate supporting formats: <atsign>:cram:<cramkey> and <atsign>:enroll:<otp>[:name:<devicename>]
- Created the ActivateParams class with parsing logic and validation
- Implemented regex-based parsing for activation strings
- Implemented help flag handling with HelpRequestedException throw-catch
- Added --keyfile option for custom atKeys file paths
- Prompts the user for a file path if a keys file already exists at the default or provided atKeys path.
- Supports verbose and debug flags

#### Issue-Keys Command:

- Generates an enrollment OTP and constructs an activation string using it. 
- Waits for the enrollment with the specific app and device name to be created; approves it once created.
- Introduced `IssueKeysParams` that has its arg parser impl and parsing
- Uses default values for namespace and appName
- Accepts user input for deviceName (fallback: noports_<OTP>), keyfile, and passphrase. Also supports verbose and debug flags.
- Support for resuming command: Checks for existing pending enrollments with given appName and deviceName. If present, it approves it and exits (so re-running the command would simply resume the flow)
- Maintains 1-hour OTP expiry with loop termination on expiration

**- How to verify it**
- Run ./buildBinaries in `sshnoports/` to generate binaries
- Run `noports issue-keys <@atsign>` to generate an enrollment string
- Copy the generated string and run `noports activate <string>` to activate
- Verify enrollment is auto-approved with correct parameters, and the keys file is generated at the user-inputted/default atKeys location

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
feat: introduce noports CLI with support for commands: 'activate' and 'issue-keys'.

srieteja committed:
feat: introduce np_activate_params class and np_activate_utils

srieteja committed:
feat: introduce np_activate_impl containing actual cram/enroll wrappers

srieteja committed:
feat: introduce np_activate_cli

srieteja committed:
build: update buildBinaries to compile noports_cli

srieteja committed:
feat: reintroduce NPActivate impl

srieteja committed:
feat: introduce NPIssueKeys impl

srieteja committed:
feat: introduce noports_cli utils/

srieteja committed:
feat: introduce CLI entrypoint 'noports' in sshnoports/bin

srieteja committed:
chore: temp pubspec overrides for at_client and onboarding_cli

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
docs: added method docs

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja committed:
build(deps): add at_auth dep in sshnoports

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
refactor: readability improvements + bug fixes + docs

srieteja committed:
build: consume at_onboarding_cli v1.14.1

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
fix: do not pass atkeys path to atClient

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+
+  const NoportsCommand(this.commandName);
+
+  static NoportsCommand fromString(String value) {
```
An enum that holds available commands and a fromString() for the same

srieteja commented on packages/dart/sshnoports/lib/src/noports_cli/activate/np_activate.dart:
```diff
@@ -0,0 +1,111 @@
+import 'package:at_auth/at_auth.dart';
+import 'package:at_onboarding_cli/at_onboarding_cli.dart';
+import 'package:sshnoports/src/noports_cli/activate/np_activate_params.dart';
+
```
activate commands implementation classes and abstractions

srieteja commented on packages/dart/sshnoports/lib/src/noports_cli/activate/np_activate_params.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:sshnoports/src/noports_cli/util/constants.dart';
+import 'package:sshnoports/src/noports_cli/util/regex.dart';
```
A params class that holds all currently necessary parameters and a fromArgs() method

srieteja commented on packages/dart/sshnoports/lib/src/noports_cli/issue_keys/np_issue_keys.dart:
```diff
@@ -0,0 +1,167 @@
+    StringBuffer cbuf = StringBuffer();
+    cbuf.append(NPIssueKeys.baseEnrollCommand
+        .replaceFirst('<atsign>', params.atsign)
+        .replaceFirst('<otp>', params.otp!));
```
baseEnrollCommand is `<atsign>:enroll:otp:<otp>`; replace placeholders with actual values

srieteja commented on packages/dart/sshnoports/lib/src/noports_cli/issue_keys/np_issue_keys.dart:
```diff
@@ -0,0 +1,167 @@
+          .fetchEnrollmentRequests(enrollmentListParams: enrollListParams);
+
+      if (enrollRequests.isEmpty) {
+        // keeps the loop active until enrollment fetched or terminated
```
Currently keeps retrying forever until successful or terminated. Should we consider having a retry limit?

gkc replied:
Nope; but we should consider having a resume if the user accidentally terminated before successful

srieteja replied:
The resume mechanism has been added through the latest commits. The mechanism works by serializing the `NPActivateParams` and then reading when available.

gkc replied:
@srieteja As I understand it, this resume is for the approving side. Maybe I'm missing something but, how is resume handled on the enrolling side? e.g. I issue the enroll: command on the enrolling side, then I hit ctrl-C. The enrollment now exists so, if I issue the same command again, I will get an exception. We discussed persisting the AtKeys file in a "pending approval" state so the enrolling side can check if that exists for the same parameters, and if so, goes into its "waiting for approval" loop without re-issuing the enroll command

srieteja replied:
Yes, this currently does not handle resume support on the enrolling side. I’ve created this [ticket](https://github.com/atsign-foundation/at_client_sdk/issues/1747)
 to track the feature in the onboarding CLI. I’ll work on that ticket next. Since the resume logic will be part of AtOnboardingService.enroll(), it will be picked up by this tool automatically once a new version of the Onboarding CLI is published.

srieteja commented on packages/dart/sshnoports/lib/src/create_at_client_cli.dart:
```diff
@@ -5,7 +5,7 @@ import 'package:path/path.dart' as path;
 
 Future<AtClient> createAtClientCli({
   required String atsign,
-  required String atKeysFilePath,
```
OnboardingCli knows the defaultAtKeysPath, and will assume that when this param is null.

srieteja commented on packages/dart/sshnoports/buildBinaries:
```diff
@@ -11,10 +11,11 @@ echo "Compiling sshnp"; dart compile exe --verbosity error bin/sshnp.dart -o bin
 echo "Compiling npp_file"; dart compile exe --verbosity error bin/npp_file.dart -o bin/npp_file &
 echo "Compiling npp_atserver"; dart compile exe --verbosity error bin/npp_atserver.dart -o bin/npp_atserver &
+echo "Compiling noports_cli"; dart compile exe --verbosity error bin/noports.dart -o bin/noports &
 echo "Compiling demo/npa_always_deny"; dart compile exe --verbosity error bin/demo/npa_always_deny.dart -o bin/demo/npa_always_deny &
```
Calling this impl `noports_cli` for now, to not confuse it with the actual noports product

xavierchanth replied:
Renaming files is more confusing, it complicated our build process.

srieteja replied:
Okay, I will change that to noports

srieteja commented on packages/dart/sshnoports/pubspec.lock:
```diff
@@ -50,7 +50,7 @@ packages:
     version: "2.13.0"
   at_auth:
-    dependency: transitive
+    dependency: "direct main"
```
Auto-generated, since at_auth was explicitly added as a dependency

xavierchanth commented on packages/dart/sshnoports/lib/src/noports_cli/activate/np_activate.dart:
```diff
@@ -0,0 +1,111 @@
+import 'package:sshnoports/src/noports_cli/activate/np_activate_params.dart';
+
+/// Handles activation of atSign accounts via CRAM authentication or APKAM enrollment
+sealed class NPActivate {
```
This doesn't need to be a sealed class. And to make it a proper facade, it should have a factory which returns an instance of the implementation.

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+    writeError('You must supply an argument string');
+    printUsage(command: command);
+    exit(1);
+  }
```
Once we have determined that args[0] is a valid command, we should remove args[0] from args so that the subsequent `wrappedMain` can concern themselves solely with the arguments they are expecting

srieteja replied:
This has been handled through the latest commits. Note that the removal was not possible as `args` is a fixed-length list. To address this, the main passes a sublist to the `wrappedMain` of individual commands, excluding the first argument.

gkc reviewed (commented):
Have left a couple of comments but looks pretty good so far; in order to complete my review I will have to try it out for real and review the code in an IDE. I will try to get to that tomorrow

srieteja committed:
Merge branch 'trunk' into noports_activate

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+
+  final String commandName;
+
+  const NoportsCommand(this.commandName);
```
I'd suggest rename NoportsCommand to NoPortsCommand

srieteja replied:
This change has been made through the latest commits.

gkc commented on packages/dart/noports_core/lib/src/activate/np_activate.dart:
```diff
@@ -0,0 +1,111 @@
+import 'dart:io';
```
I'd suggest you move all of this code into the noports_core package so it can be reused elsewhere (since we don't publish the sshnoports package

xavierchanth replied:
+1, we will need some of this on the flutter side.

srieteja replied:
I agree. But since a review is already in progress here, I think waiting until the review is complete and then relocating the code would be efficient.

srieteja replied:
I have moved to code to noports_core/activate through [8c7f4a3](https://github.com/atsign-foundation/noports/pull/2304/commits/8c7f4a35583c142ed27216b4d84cc84dbe0c5ffe).

gkc commented on packages/dart/sshnoports/lib/src/noports_cli/activate/np_activate.dart:
```diff
@@ -0,0 +1,111 @@
+import 'package:sshnoports/src/noports_cli/activate/np_activate_params.dart';
+
+/// Handles activation of atSign accounts via CRAM authentication or APKAM enrollment
+sealed class NPActivate {
```
What's the reason for this being a sealed class?

srieteja replied:
I was going for abstraction in the earlier design and I had to prevent the abstract class from being implemented externally. Since the design has moved away from that pattern, there is no longer a need for it to be a sealed class.

gkc commented on packages/dart/sshnoports/lib/src/noports_cli/activate/np_activate.dart:
```diff
@@ -0,0 +1,111 @@
+/// Determines the type of activation: CRAM or APKAM enrollment
+enum NPActivateType { cram, enroll }
+
+class NPActivateImpl implements NPActivate {
```
Currently the only way of getting an instance of NPActivate is to construct an NPActivateImpl. What is the thinking behind this design?

srieteja replied:
In the intial design, I introduced abstraction to guide callers on which methods were intended for use; similar to the current `AtClient` and `AtOnboardingService` patterns. Although after further consideration and discussion on this PR, I have moved away from that pattern and adoped a 'facade with a factory' pattern which fits the usecase better.

gkc commented on packages/dart/sshnoports/lib/src/noports_cli/activate/np_activate_params.dart:
```diff
@@ -0,0 +1,69 @@
+import 'package:sshnoports/src/noports_cli/util/constants.dart';
+import 'package:sshnoports/src/noports_cli/util/regex.dart';
+
+class NoportsParams {
```
Shouldn't be called NoportsParams; the `noports` binary is ultimately intended as a wrapper which starts up any of the noports programs. It has no parameters of its own. Since this class is specifically about the `noports activate` command, it should be named accordingly.

srieteja replied:
This class has been renamed to ActivateParams in the latest commits.

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+}
+
+final NPActivate npActivate = NPActivateImpl();
+final NPIssueKeys issueKeys = NPIssueKeysImpl();
```
Why are these instances in the top-level scope rather than local within the `main` function?

srieteja replied:
The first draft has other functions that needed access to these instances. I have now moved this into main() as they're no longer needed in that scope.

gkc commented on packages/dart/sshnoports/lib/src/noports_cli/util/usage_messages.dart:
```diff
@@ -0,0 +1,36 @@
+class UsageMessages {
+  static final mainMenu = '''
+  
+Usage: noports <command> <argument_string>
```
`<argument_string>` suggests there is only one argument string. But it's actually going to be `noports <command> [arg1, arg2, ...]>`

xavierchanth replied:
I suggest using `brew --help` / `brew commands` / `brew help [COMMAND]` as an example for a good experience for a router/wrapper cli app

srieteja replied:
I changed the <argument_string> to [options] for a broader scope. I have also updated the help messages to be more user-friendly.

gkc commented on packages/dart/sshnoports/lib/src/noports_cli/util/np_utils.dart:
```diff
@@ -0,0 +1,31 @@
+
+import 'package:chalkdart/chalk.dart';
+
+void writeError(String message) {
```
Why are we doing all of this instead of just using an AtSignLogger?

srieteja replied:
I have introduced a custom logging handler as we discussed.

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+
+  final command = NoportsCommand.fromString(args[0]);
+  if (args.length == 1) {
+    writeError('You must supply an argument string');
```
You should leave this up to the parser for the actual command.

srieteja replied:
Has been addressed through the latest commits.

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+    exit(1);
+  }
+
+  final command = NoportsCommand.fromString(args[0]);
```
If you run `noports foo` then you get an ugly stack trace. Need to wrap in its own try-catch block like
```
  final NoportsCommand command;
  try {
    command = NoportsCommand.fromString(args[0]);
  } catch (e) {
    writeError(e.toString());
    printUsage();
    exit(1);
  }
```

srieteja replied:
Most edge cases have been identified and are being appropriately handled. This snippet has also been included to handle potential exceptions from NoPortsCommand.fromString().

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,88 @@
+    /// ToDo: parse exception message to remove unnecessary Exception prefixes
+    writeError(e.toString());
+    stderr.writeln();
+    stderr.writeln('Please try again or contact support@atsign.com');
```
You do not need to output this message.

srieteja replied:
This message has been removed in the latest commits.

srieteja committed:
feat: introduce cli_logging_handler.dart

srieteja committed:
feat: refactor utils/

srieteja committed:
feat: support for stateFile to allow resuming issue-keys command

srieteja committed:
feat: support for stateFile to allow resuming issue-keys command

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

gkc commented on packages/dart/noports_core/lib/src/common/cli_logging_handler.dart:
```diff
@@ -0,0 +1,33 @@
+        return chalk.red('ERR');
+      case Level.INFO:
+      default:
+        return chalk.blueBright('INFO');
```
I'd rather you had explicit handling here for FINER and FINEST

srieteja replied:
Explicit handling for FINER and FINEST has been added in the latest commit.

xavierchanth replied:
Why not upstream this to atlogger hidden behind a boolean switch?

srieteja replied:
Will remove CLILoggingHandler from here once a new version of at_utils is published.

gkc replied:
OK

srieteja committed:
feat: refactoring for readability

srieteja commented on packages/dart/noports_core/lib/src/commands/issue_keys/issue_keys.dart:
```diff
@@ -0,0 +1,251 @@
+      ..enrollmentListFilter = [EnrollmentStatus.pending];
+
+    logger.info('Listening...');
+
```
I've added a retry limit. This is needed because the OTP has a TTL — default is about 5 minutes. Right now, if the user runs issue-keys and then runs activate after the OTP expires, we end up in an endless wait: issue-keys keeps waiting forever for an enrollment that will never show up, and activate can’t create the enrollment because the OTP is already expired. By adding a retry limit, we can control when the waiting stops, handle the timeout cleanly, and tell the user exactly what's going on.

Also note that the `maxRetries` is calculated by dividing `otpExpiry` by `retryInterval`(both in seconds). This syncs the `retries` to the `otpExpiry` .

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
fix: help bug

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

gkc committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
refactor: move noports_cli from sshnoports -> noports_core

srieteja committed:
build(deps): update at_onboarding_cli to v1.14.2

srieteja committed:
chore: dart formatter code cleanup

srieteja committed:
build(deps): remove at_auth dep from sshnoports

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
build(deps): update pubspec.lock in np_admin

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

gkc committed:
Merge branch 'trunk' into noports_activate

gkc committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
feat: Introduce separate parser for issue-keys and consume it

srieteja committed:
feat: activate command uses a hybrid parser. relocated to noports_core/src/commands/

srieteja committed:
feat: move utils to noports_core/src/commands/util/

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
feat: introduce HelpRequestedException use it

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja commented:
Changes in the last batch of commits:

1. Moved the code to noports_core/src/commands/
2. Created new exception `HelpRequestedException` that is thrown when help is requested, help messages displayed when caught.
3. Created a separate arg parser + params class for IssueKeys: `IssueKeysParams`
4. Modified the existing activate params class to a hybrid one: Regex based reading of params from activation_string, then an arg parser for the rest.
5. Removed the state file based resume mechanism; replaced with a check for matching pending enrollments before the usual flow. 
6. Updated help messages to be more descriptive and included examples

srieteja commented on packages/dart/noports_core/lib/src/commands/activate/activate_impl.dart:
```diff
@@ -0,0 +1,146 @@
+    final String? keysFilePath;
+
+    if (_params.atKeysFilePath == null) {
+      // Use the default path from onboarding service preferences
```
Workaround to avoid conflicts with creating atKeys files. Currently fetches the defaultPath from OnboardingServiceImpl.

Changes required: OnboardingService should throw a dedicated exception for key file name collisions, which can then be caught and handled.

gkc replied:
Need to address this issue fully in this PR. Find the places that MUST have file existence checking (onboard, enroll), ensure the checking always happens there. Remove file existence checking from elsewhere and verify that all the ways we create atKeys files do end up having a file existence check so we don't ever overwrite keys files. As a final safeguard, put a check into the place we finally write an AtKeys file and do not permit overwriting.

gkc replied:
@srieteja are you including the conflict handling in this PR or in a different PR?

xavierchanth commented on packages/dart/noports_core/lib/src/commands/activate/activate_params.dart:
```diff
@@ -0,0 +1,134 @@
+      );
+    }
+
+    final activationString = results.rest.first;
```
Why .first instead of .single?

srieteja replied:
.first keeps the cli lenient by avoiding errors when extra values are passed. But .single seems clearer and  more straighforward, so I've replaced .first with .single through [90a00f6](https://github.com/atsign-foundation/noports/pull/2304/commits/90a00f64245a572d0c4a5daca989208fb03e2ec9)

xavierchanth commented on packages/dart/noports_core/lib/src/commands/activate/activate_params.dart:
```diff
@@ -0,0 +1,134 @@
+      );
+  }
+
+  static String? _parseAtsign(String input, ActivateType type) {
```
at_commons has an AtSign type which can be used now. This type calls the equivalent of AtUtils.FixAtsign under the hood.

srieteja replied:
Addressed through [90a00f6](https://github.com/atsign-foundation/noports/pull/2304/commits/90a00f64245a572d0c4a5daca989208fb03e2ec9)

srieteja commented on packages/dart/noports_core/lib/src/commands/issue_keys/issue_keys.dart:
```diff
@@ -0,0 +1,225 @@
+  /// Returns: 0 on success, 1 on failure
+  Future<int> wrappedMain() async {
+    await _initAtClient();
+
```
If existing matching enrolment present, approves it

xavierchanth commented on packages/dart/noports_core/lib/src/commands/activate/activate_impl.dart:
```diff
@@ -0,0 +1,146 @@
+  final logger = AtSignLogger('Activate', loggingHandler: CLILoggingHandler())
+    ..level = 'info';
+
+  Activate._(this._onboardingService, this._params);
```
Nit: the strongly constructor shouldn't be private, most non-binary consumers of this api will prefer this over the factory constructor.

srieteja replied:
Constructor made public through [90a00f6](https://github.com/atsign-foundation/noports/pull/2304/commits/90a00f64245a572d0c4a5daca989208fb03e2ec9)

xavierchanth commented on packages/dart/noports_core/lib/src/commands/issue_keys/issue_keys_impl.dart:
```diff
@@ -0,0 +1,225 @@
+    loggingHandler: CLILoggingHandler(),
+  )..level = 'info';
+
+  IssueKeys._(this._params);
```
Same here.

srieteja replied:
Constructor made public through [90a00f6](https://github.com/atsign-foundation/noports/pull/2304/commits/90a00f64245a572d0c4a5daca989208fb03e2ec9)

srieteja committed:
fix: changes from self review

xavierchanth commented on packages/dart/noports_core/lib/src/commands/utils/regex.dart:
```diff
@@ -0,0 +1,21 @@
+
+  // Enrollment: <atsign>:enroll:otp:<otp>[:name:<device>:keyfile:<path>]
+  static final enroll = RegExp(
+    r'^(?<atsign>[^:]+):enroll:otp:(?<otp>[A-Za-z0-9]{6})'
```
This is ordered, which is fine simple is best. Just make sure the Desktop app respects the same order  when it gets implemented there.

Ideally the app consumes the same classes.

srieteja replied:
This was left behind. The keys file is not accepted through the activation_string anymore; it can be passed as a separate arg with `-k (or) -key-file`. I'll make sure the desktop app works as expected.

xavierchanth commented on packages/dart/noports_core/lib/src/commands/utils/usage_messages.dart:
```diff
@@ -0,0 +1,74 @@
+${chalk.bold('Usage:')} noports activate ${chalk.cyan('<activation_string>')} [options]
+
+${chalk.bold('Options:')}
+  -k, --keyfile <path>        Path to save atKeys file
```
Looks like the help wasn't updated

srieteja replied:
I did update it earlier, I also reworked them again through ([8a4a9d7](https://github.com/atsign-foundation/noports/pull/2304/commits/8a4a9d7ed63ad7593a82669e699a1dbda2b53f6f)).

xavierchanth commented on packages/dart/noports_core/lib/src/common/help_requested_exception.dart:
```diff
@@ -0,0 +1,3 @@
+class HelpRequestedException implements Exception{
```
Your Pr needs formatting

srieteja replied:
addressed in my last commit

xavierchanth reviewed (dismissed):
Minor changes and some questions.

JeremyTubongbanua commented:
Feedback:

- `--version` or show version in usage instructions

So there's 3 functions to the NoPorts CLI that I'm aware of

- `noports activate` (CRAM Onboarding)
- `noports issue-keys` (OTP and Auto Approve)
- `noports enroll` (we want this to be a new command, separate from activate)

Right now 1 and 3 are in the same `activate` command, but I think it should be separate commands

srieteja committed:
fix: constructors made public + minor formatting

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
docs: updated usage messages + minor fixes

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

xavierchanth commented:
> Right now 1 and 3 are in the same `activate` command, but I think it should be separate commands

Then it no longer solves the original goal.

The original goal is to create a string format that encodes all of the information required to transmit an atSign between machines for use with NoPorts. As soon as you separate the commands, or add options and flags, you have now coupled the use-case to the CLI and it is no longer compatible with the desktop app.

JeremyTubongbanua commented:
> > Right now 1 and 3 are in the same `activate` command, but I think it should be separate commands
> 
> Then it no longer solves the original goal.
> 
> The original goal is to create a string format that encodes all of the information required to transmit an atSign between machines for use with NoPorts. As soon as you separate the commands, or add options and flags, you have now coupled the use-case to the CLI and it is no longer compatible with the desktop app.

Makes sense,

It was just some feedback I had during my 30min call with Srie. If there was an original vision that I wasn't aware of, then we can ignore my suggestion

xavierchanth commented:
> > > Right now 1 and 3 are in the same `activate` command, but I think it should be separate commands
> > 
> > 
> > Then it no longer solves the original goal.
> > The original goal is to create a string format that encodes all of the information required to transmit an atSign between machines for use with NoPorts. As soon as you separate the commands, or add options and flags, you have now coupled the use-case to the CLI and it is no longer compatible with the desktop app.
> 
> Makes sense,
> 
> It was just some feedback I had during my 30min call with Srie. If there was an original vision that I wasn't aware of, then we can ignore my suggestion

This is covered in my onboarding document.

srieteja committed:
Merge branch 'trunk' into noports_activate

gkc committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
chore: minor tweaks and cleanup

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja committed:
Merge branch 'trunk' into noports_activate

gkc commented on packages/dart/noports_core/lib/src/commands/activate/activate_impl.dart:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    logger.shout('KeysFile exists at $keysFilePath');
+    _params.atKeysFilePath = promptUser(
```
what happens if the user puts the same path in again?

srieteja replied:
This would fail again. As discussed previously, I’m working on a better solution in AtOnboardingCli. Once that work is in place, this check will be updated to use the collision handler.

gkc commented on packages/dart/noports_core/lib/src/commands/activate/activate_params.dart:
```diff
@@ -0,0 +1,131 @@
+      );
+  }
+
+  static String? _parseAtsign(String input, ActivateType type) {
```
Return type should be `Atsign?` not `String?`

srieteja replied:
This has been fixed through b548404653fd76f9109b95787139125c51b39ad7

gkc commented on packages/dart/noports_core/lib/src/commands/activate/activate_params.dart:
```diff
@@ -0,0 +1,131 @@
+    return match?.namedGroup(ActivateRegexGroups.deviceName);
+  }
+
+  Map<String, String?> toJson() {
```
Why is `type` not part of the json?

gkc replied:
You are missing a test somewhere to ensure this kind of problem doesn't happen

gkc replied:
after further discussion we concluded that toJson is no longer needed and can be removed from this class

srieteja replied:
The toJson method has now been removed.

gkc commented on packages/dart/noports_core/lib/src/commands/issue_keys/issue_keys.dart:
```diff
@@ -0,0 +1,211 @@
+    if (existingEnrollment != null) {
+      await _approveEnrollment(existingEnrollment);
+      return 0;
+    }
```
personal preference, I would prefer the rest of this code were in an `else` block

srieteja replied:
I've put the rest of the code in a else block through b548404653fd76f9109b95787139125c51b39ad7.

gkc commented on packages/dart/noports_core/lib/src/commands/issue_keys/issue_keys.dart:
```diff
@@ -0,0 +1,211 @@
+
+    if (results.isNotEmpty) {
+      logger.info('Enrollment found with id: ${results.first.enrollmentId}');
+      return results.first;
```
Should we be asserting there be just one? Or at least an explanation that we can only expect at most one result.

srieteja replied:
Yes, I've added the check to ensure there can only be one enrollment request for a given device name. In the case of multiple enrollments, an AtEnrollmentException is now thrown.

gkc commented on packages/dart/sshnoports/bin/noports.dart:
```diff
@@ -0,0 +1,99 @@
+  displayBanner();
+  AtSignLogger.root_level = 'severe';
+  return AtSignLogger('NoPorts', loggingHandler: CLILoggingHandler())
+    ..level = 'info';
```
No way for the user to change the log level?

gkc replied:
add `--verbose | -v` and `--debug` as we have for sshnpd

srieteja replied:
Verbose and debug args have been added through b548404653fd76f9109b95787139125c51b39ad7.

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
chore: remove cli_logging_handler, consume from at_utils

srieteja committed:
feat: improve atKeys file collision logic

srieteja committed:
feat: changes to support testing

srieteja committed:
test: add unit tests

srieteja committed:
chore: other minor changes

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja committed:
test: more tests

srieteja committed:
fix: git actions failing

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
refactor: rename activate_impl.dart -> activate.dart and issue_keys_impl.dart -> issue_keys.dart

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
test: more unit tests + required impl refactor

srieteja committed:
Merge remote-tracking branch 'origin/noports_activate' into noports_activate

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
test: minor test cleanup

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
Merge branch 'trunk' into noports_activate

srieteja committed:
test: fix unit failing tests (Atsign type mismatch)

srieteja committed:
Merge branch 'trunk' into noports_activate

gkc committed:
Merge branch 'trunk' into noports_activate

gkc reviewed (approved):
LGTM. Please follow up with a PR to change `.github/workflows/multibuild.yaml` to add in the noports binary into our releases. Please also update the `buildArchive` script in that PR.
